purchase_discount/models/purchase_order.py

```diff
@@ -0,0 +1,61 @@
+# -*- coding: utf-8 -*-
+# Copyright 2004-2009 Tiny SPRL (<http://tiny.be>).
+# Copyright 2016 ACSONE SA/NV (<http://acsone.eu>)
+# Copyright 2015-2017 Tecnativa - Pedro M. Baeza
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+import odoo.addons.decimal_precision as dp
+
+
+class PurchaseOrderLine(models.Model):
+    _inherit = "purchase.order.line"
+
+    @api.depends('discount')
+    def _compute_amount(self):
+        super(PurchaseOrderLine, self)._compute_amount()
+
+    def _prepare_compute_all_values(self):
+        vals = super(PurchaseOrderLine, self)._prepare_compute_all_values()
+        # This is always executed for allowing other modules to use this
+        # with different conditions than discount != 0
+        vals.update({'price_unit': self._get_discounted_price_unit()})
+        return vals
+
+    discount = fields.Float(
+        string='Discount (%)', digits=dp.get_precision('Discount'),
+    )
+
+    _sql_constraints = [
+        ('discount_limit', 'CHECK (discount <= 100.0)',
+         'Discount must be lower than 100%.'),
+    ]
+
+    def _get_discounted_price_unit(self):
+        """Inheritable method for getting the unit price after applying
+        discount(s).
+
+        :rtype: float
+        :return: Unit price after discount(s).
+        """
+        self.ensure_one()
+        if self.discount:
+            return self.price_unit * (1 - self.discount / 100)
+        return self.price_unit
+
+    @api.multi
+    def _get_stock_move_price_unit(self):
+        """Get correct price with discount replacing current price_unit
+        value before calling super and restoring it later for assuring
+        maximum inheritability.
+        """
+        price_unit = False
+        price = self._get_discounted_price_unit()
+        if price != self.price_unit:
+            # Only change value if it's different
+            price_unit = self.price_unit
+            self.price_unit = price
+        price = super(PurchaseOrderLine, self)._get_stock_move_price_unit()
```

Comment:
Maybe it is possible to call super at the beginning and then check the discount and change the value is needed, don't you think?

Reply:
@aheficent  Not sure. But looking at the code I am feeling that at first price_unit is changed with the discounted amount (unit_price - discount) and then calling super so that everything in the super method(s) will get calculated based on discounted unit_price and then changing the unit_price back to original.

Reply:
But the original method is only getting the price unit. Then you can change it or not. Anyway it is not blocking.